### PR TITLE
Support for URL Encoded BodyParam

### LINKED ISSUE #20350: Support BodyParams using application/x-www-form-urlencoded
Currently, `application/x-www-form-urlencoded` body types only are generated properly when each form segment is declared using `FormPart`.

```java
public Response<Void> restApi(@FormPart("id") long id, @FormPart("name") String name, @FormPart("age") int age)
```

But there may be cases, either written manually or code generated, that want to use an object type for the body.

```java
public Response<Void> restApi(@BodyParam("application/x-www-form-urlencoded") Person person);

public class Person {
    public long id;
    public String name;
    public int age;
}
```

Support for this should be enabled either with better handling for `BodyParam` when `Content-Type` is `application/x-www-form-urlencoded` or by having code generation use `FormPart`, or another new annotation such as `FormBody`, in this case.

### PR DESCRIPTION
**Pull request by alzimmermsft**
Fixes #20350 

This PR enables `BodyParam` to handle cases where the `Content-Type` is `application/x-www-form-urlencoded`. When setting preparing the request body the body object will be converted into a URL encoded string.

### CODE REVIEW

File: sdk/core/azure-core/src/main/java/com/azure/core/http/rest/SwaggerMethodParser.java [hunk 1]
```diff
@@ -201,11 +167,17 @@
...
                     bodyContentMethodParameterIndex = parameterIndex;
                     bodyContentType = bodyParamAnnotation.value();
-                    bodyJavaType = swaggerMethod.getGenericParameterTypes()[parameterIndex];
+
+                    // If the body Content-Type is "application/x-www-form-urlencoded" the real body type is a String.
+                    bodyJavaType = (ContentType.APPLICATION_X_WWW_FORM_URLENCODED.equals(bodyContentType))
+                        ? String.class : swaggerMethod.getGenericParameterTypes()[parameterIndex];
                 } else if (annotationType.equals(FormParam.class)) {
                     final FormParam formParamAnnotation = (FormParam) annotation;
-                    formSubstitutions.add(new Substitution(formParamAnnotation.value(), parameterIndex,
```
Comment (pallavit) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/rest/SwaggerMethodParser.java:207]: >formSubstitutions [](start = 20, length = 17)

when will autorest generate @FormParam vs @BodyParam for APPLICATION_X_WWW_FORM_URLENCODED

Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/rest/SwaggerMethodParser.java:207]: @jianghaolu, @srnagar ?

Comment (jianghaolu) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/rest/SwaggerMethodParser.java:207]: Currently AutoRest doesn't support generating `FormParam` due to lack of support in azure-core. You tell us what to generate for the following scenarios and we can generate as desired:

1. Upload file through multipart form-data. See [swagger](https://github.com/Azure/autorest.testserver/blob/master/swagger/body-formdata.json#L42). Consumes `multipart/form-data`. I expect this to be a `FormParam`.
2. Upload file through body. See [swagger](https://github.com/Azure/autorest.testserver/blob/master/swagger/body-formdata.json#L79). Consumes `application/octet-stream`. I expect this to be a `BodyParam`.
3. Upload file through form url encoded form-data. See [swagger](https://github.com/Azure/autorest.testserver/blob/master/swagger/body-formdata-urlencoded.json#L46). Consumes `application/x-www-form-urlencoded`. I prefer this to be a `FormParam` but then how do we differentiate with multi-part? I'm okay with either on this.

Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/rest/SwaggerMethodParser.java:207]: I would use `FormPart` for `multipart/form-data` given that we've already tied `FormParam` to URL encoded

**Comment by alzimmermsft** [comment]
Closing as code generation resolves this issue

**Pull request closed without merging**